Connectors/Rhino/Speckle.Connectors.GrasshopperShared/Parameters/SpeckleBlockDefinitionWrapper.cs

```diff
@@ -0,0 +1,512 @@
+        };
+        return true;
+
+      case InstanceDefinition rhinoInstanceDef:
+        return CastFromRhinoInstanceDefinition(rhinoInstanceDef);
+    }
+
+    // Handle Rhino 8 Model Objects
+    return CastFromModelObject(source);
+  }
+
+#if !RHINO8_OR_GREATER
+  private bool CastFromModelObject(object _) => false;
+
+  private bool CastToModelObject<T>(ref T _) => false;
+#endif
+
+  public override bool CastTo<T>(ref T target)
+  {
+    var type = typeof(T);
+
+    if (type == typeof(InstanceDefinitionProxy))
+    {
+      target = (T)(object)Value.InstanceDefinitionProxy;
+      return true;
+    }
+
+    if (type == typeof(InstanceDefinition))
+    {
+      return CastToRhinoInstanceDefinition(ref target);
+    }
+
+    return CastToModelObject(ref target);
+  }
+
+  /// <summary>
+  /// Converts a Rhino <see cref="InstanceDefinition"/> to a <see cref="SpeckleBlockDefinitionWrapper"/>.
+  /// Takes all objects from the instance definition and converts them to <see cref="SpeckleObjectWrapper"/>s.
+  /// </summary>
+  private bool CastFromRhinoInstanceDefinition(InstanceDefinition instanceDef)
+  {
+    try
+    {
+      var objects = new List<SpeckleObjectWrapper>();
+      var objectIds = new List<string>();
+      var units = RhinoDoc.ActiveDoc?.ModelUnitSystem.ToSpeckleString() ?? "none";
+
+      var rhinoObjects = instanceDef.GetObjects(); // Get the objects that DEFINE the block, not all instances of it
+      foreach (var rhinoObj in rhinoObjects)
+      {
+        if (rhinoObj?.Geometry != null)
+        {
+          Base converted; // SpeckleConversionContext.ConvertToSpeckle() is for basic geometry, we may have nested instances
+
+          if (rhinoObj.Geometry is InstanceReferenceGeometry instanceRef) // nested!
+          {
+            // get nested instance definition for name lookup
+            var nestedInstanceDef = RhinoDoc.ActiveDoc?.InstanceDefinitions.FindId(instanceRef.ParentIdefId);
+            string definitionName = nestedInstanceDef?.Name ?? instanceRef.ParentIdefId.ToString();
+
+            // create an InstanceProxy for nested blocks using correct properties
+            var instanceProxy = new InstanceProxy
+            {
+              definitionId = instanceRef.ParentIdefId.ToString(),
+              transform = GrasshopperHelpers.TransformToMatrix(instanceRef.Xform, units),
+              units = units,
+              maxDepth = 1,
+              applicationId = rhinoObj.Id.ToString()
+            };
+            instanceProxy["definitionName"] = definitionName;
+            converted = instanceProxy;
+          }
+          else // regular geometry
+          {
+            converted = SpeckleConversionContext.ConvertToSpeckle(rhinoObj.Geometry);
+          }
+
+          converted[Constants.NAME_PROP] = rhinoObj.Name ?? "";
+          converted.applicationId = rhinoObj.Id.ToString();
+
+          var objWrapper = new SpeckleObjectWrapper()
```

Comment:
We're handling conversions with casting -> this can be changes to creating a new Goo() and using Goo.CastFrom() to deal with all conversions from rhino objects/ geometryBase/instances